src/KingSystem/Damage/dmgDamageManagerBase.cpp

```diff
@@ -0,0 +1,271 @@
+            DamageCallback* prev;
+            while (next) {
+                prev = next;
+                next = next->mNext;
+            };
+
+            prev->mNext = callback;
+            callback->mPrev = prev;
+        } else {
+            mCallbacks[eventId] = callback;
+        }
+
+        callback->mDamageManager = this;
+        callback->mEventId = eventId;
+    }
+}
+
+void DamageManagerBase::removeDamageCallback(DamageCallback* callback) {
+    if (!mCallbacks.isBufferReady() || callback->mDamageManager != this) {
+        return;
+    }
+
+    u32 eventId = callback->mEventId;
+    DamageCallback* currentCallback = mCallbacks[eventId];
+    if (!currentCallback) {
+        if (mActor) {
+            // Logging about trying to remove missing callback?
+            mActor->nullsub_4649();
+        }
+
+        callback->mDamageManager = NULL;
+        callback->mEventId = -1;
+        asm("");  // Stop optimizing with the other clear below
+        return;
+    }
+
+    if (currentCallback == callback) {
+        mCallbacks[eventId] = callback->mNext;
+        currentCallback = mCallbacks[eventId];
+        if (currentCallback) {
+            currentCallback->mPrev = NULL;
+        }
+        asm("");  // Stop re-using variables, generating an extra register
+    } else {
+        do {
+            if (currentCallback == callback) {
+                DamageCallback* prev = callback->mPrev;
+                if (prev) {
+                    prev->mNext = callback->mNext;
+                }
+                DamageCallback* next = callback->mNext;
+                if (next) {
+                    next->mPrev = callback->mPrev;
+                }
+            }
+            currentCallback = currentCallback->mNext;
+        } while (currentCallback);
+    }
+    callback->mNext = NULL;
+    asm("");  // Stop combining the mNext and mDamageManager
+    callback->mDamageManager = NULL;
+    callback->mEventId = -1;
+    callback->mPrev = NULL;
+}
+
+bool DamageManagerBase::applyDamage(s32& life) {
+    ksys::res::GParamList* paramsList;  // x8
+    paramsList = mActor->mActorParam->gParamList;
+
+    const ksys::res::GParamListObjectGeneral& params = paramsList->getGeneral();
+    if (params.mIsLifeInfinite.ref()) {
+        // Since life is infinite, we don't need to modify the damage or life.
+        // But we still call the "callback" as if damage was done.
+        onApplyDamage();
+        return false;
+    }
+
+    s32 damage = getDamage();
+    if (damage >= 1) {
+        if (isPlayerClass(mActor)) {
+            if (getAttacker()->hasProc()) {
+                tryBuffDamage(damage);
+            }
+        }
+    }
+
+    tryApplyDamageRecovery(damage);
+
+    // Actually deal damage
+    onApplyDamage();
+    f32 oldLife = life;  // Random float for some reason
+    // Force positive result(Any negative result right shifted by 31 = -1, positive result = 0)
+    life = (life - damage) & ~((life - damage) >> 0x1F);
```

Comment:
Does this match with `std::max(0, life - damage)`?

Reply:
Yes, it did :)